Fix the non-default constructor mechanism of bytecode recording

gastaldi opened the pull request:
- Fix the non-default constructor mechanism of bytecode recording
- Fixes #31153
- Optimize bytecode recording test performance

gastaldi committed:
Include test case for #31153

Fix the test case for #31153

gastaldi committed:
Optimize bytecode recording test performance

- Change creation of a large collection in `BytecodeRecorderTestCase.java` from using a loop to using `Collections.nCopies`

Ladicek committed:
Fix the non-default constructor mechanism of bytecode recording

Bytecode recorders attempt to match constructor parameters to properties
and skip setting properties that were initialized by the constructor.
The assumption is that if the constructor has a parameter with the same
name as the property, the property is set by the constructor. Otherwise
it needs to be set using a setter.

This matching was not implemented for the non-default constructor mechanism.
This commit fixes that.

gastaldi commented:
Thanks to @Ladicek and @geoand who helped with the investigation!

geoand commented:
I didn't do anything, you guys did all the work :)

gastaldi commented:
@geoand well, you approved this PR, so that's something 😉

gastaldi commented:
@Ladicek looks like ArC didn't like that change:

```
Could not find parameters for constructor public io.smallrye.config.ConfigMappings$ConfigClassWithPrefix(java.lang.Class,java.lang.String) could not read field values [arg1, arg0]
```

gastaldi commented on core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java:
```diff
@@ -1173,6 +1173,13 @@ public void prepare(MethodContext context) {
+            if (nonDefaultConstructorHolder.constructor.getParameterCount() > 0) {
+                Parameter[] parameters = nonDefaultConstructorHolder.constructor.getParameters();
+                for (int i = 0; i < parameters.length; ++i) {
+                    String name = parameters[i].getName();
```
I think we need to check if `java.lang.reflect.Parameter#isNamePresent` here, WDYT @Ladicek ?

Ladicek replied:
I was wondering why the existing code (which I merely copied) doesn't check if parameter names exist, but it seemed to work... :shrug:

Yeah, I agree we shouldn't put the synthetic parameter names `argN` to the map. That, together with how the map is used (matching parameters are removed from it, and then the map is checked if some parameters remain), should fix the issue here.

gastaldi replied:
@Ladicek mind pushing the fix to this PR?

Ladicek replied:
Pushed a commit, let's see how that goes :-)

Ladicek committed:
Fix bytecode recorder matching of properties to constructor parameters

Bytecode recorders attempt to match properties to constructor parameters
by the property/parameter name. However, parameters do not always have
to have their names available. This commit fixes the bytecode recorder
matching mechanism to avoid parameters that do not have an available name.

gastaldi commented:
Test failures are unrelated and happen in current `main` branch

gsmet commented:
If we want it backported, we will need specific PRs. It doesn't apply cleanly and I prefer having the author fix the conflicts given it's a sensitive matter.

gastaldi commented:
I'll refrain from backporting this to 2.13 because it involves changes that would require a deeper analysis (like the introduction of `RecordingAnnotationsUtil`)
